### DISCUSSION TITLE
Provide contextual type hinting for protected members

### DISCUSSION THREAD

**Discussion by evanmags** [microsoft/pylance-release#4439] [Enhancement]
## Environment data

-   Language Server version: 2023.5.40
-   OS and version: MacOS Ventura 13.3.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.11 through 3.11.3 (likely all, these are just the ones I use)

## Code Snippet

```python
class PrivateEx:
    def __private_method(self):
        print("you found me! 🍁💩")

    def __public_dunder__(self):
        print("I wasn't hidden!")


p = PrivateEx()

try:
    p.__private_method()
except AttributeError as ae:
    print(ae)

p.__public_dunder__()

p._PrivateEx__private_method()
```

## Repro Steps

The code snippet above is executable as is to reproduce locally.

## Expected behavior

- when referencing a `Self` type, protected methods should be displayed in intellisense by their defined name.
- when referencing an instance, protected methods should either not be displayed at all or displayed as `_{class_name}{method_name}`in intellisense.
- dunder methods should not be effected.

## Actual behavior

- protected methods are shown in intellisense regardless of context, this can lead to unexpected autocomplete suggestions or runtime exceptions.

## Logs

n/a

**Comment by erictraut** [comment]
A couple of points:
1. Names within a class scope that begin with a double underscore (but are not dunders) are officially referred to as "private members" in the [Python documentation](https://docs.python.org/3/tutorial/classes.html#private-variables). That's probably what you meant when you said "protected members".
2. Private member access is allowed anywhere within the class implementation, not only when referenced via `self` variables (as per your first bullet point above).
3. Referencing private members by the mangled name outside of the class is not recommended, so I don't think that should ever be suggested by the completion provider (as per your second bullet point above).

I understand that the request is for a change to the completion provider (the component that provides Intellisense suggestions), and I'll leave it to the pylance team to respond to that. I'll note that there's a pyright diagnostic check that you can enable that will flag the use of private members outside of the class. It's called `reportPrivateUsage`. You can enable it globally for your project or for specific files (by adding a `# pyright: reportPrivateUsage=true` comment at the top of your file). This will alert you to potential runtime exceptions at edit time.

**Comment by debonte** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
